frame/support/src/traits.rs

```diff
@@ -132,3 +139,66 @@ impl<Imbalance: TryDrop> OnUnbalancedKton<Imbalance> for () {
 		drop(amount);
 	}
 }
+
+// A regulator to adjust relay args for a specific chain
+// Implement this in runtime's impls
+pub trait AdjustableRelayerGame {
+	type Moment;
+	type Balance;
+	type TcBlockNumber;
+
+	fn challenge_time(round: Round) -> Self::Moment;
+
+	fn round_from_chain_len(chain_len: u64) -> Round;
+
+	fn chain_len_from_round(round: Round) -> u64;
+
+	fn update_samples(round: Round, samples: &mut Vec<Self::TcBlockNumber>);
+
+	fn estimate_bond(round: Round, proposals_count: u64) -> Self::Balance;
+}
+
+/// Implement this for target chain's relay module's
+/// to expose some necessary APIs for relayer game
+pub trait Relayable {
+	type TcBlockNumber: Clone + Copy + Debug + Default + AtLeast32Bit + FullCodec;
+	type TcHeaderHash: Clone + Debug + Default + PartialEq + FullCodec;
+	type TcHeaderMMR: Clone + Debug + Default + PartialEq + FullCodec;
+
+	/// The latest finalize block's header's record id in darwinia
+	fn last_confirmed() -> Self::TcBlockNumber;
+
+	/// Check the header if it's already existed
+	fn header_existed(block_number: Self::TcBlockNumber) -> bool;
+
+	/// Verify the codec style header thing
+	fn verify_raw_header_thing(
+		raw_header_thing: RawHeaderThing,
+	) -> Result<
+		TcHeaderBrief<Self::TcBlockNumber, Self::TcHeaderHash, Self::TcHeaderMMR>,
+		DispatchError,
+	>;
+
+	/// Verify the codec style header thing chain
+	fn verify_raw_header_thing_chain(
```

Comment:
What's the difference between ```verify_raw_header_thing_chain``` and ```verify_raw_header_thing ```?

Reply:
Implement `verify_raw_header_thing` to define how to verify a single header.

Once you implement the `verify_raw_header_thing`, `verify_raw_header_thing_chain` will be implemented automatically which use for verifying a whole chain. Or you can implement `verify_raw_header_thing_chain` again that you can customize this function (e.g verify continuous).

---

Why `verify_raw_header_thing`?

Sometimes we don't need to verify the whole chain to make a checking. (e.g something wrong with `game_id` return error)